src/generic_solver.cpp

```diff
@@ -81,16 +101,189 @@ GenericSolver::~GenericSolver() {
   delete read_buf;
   delete term_counter;
   // close the solver process
-  // TODO uncomment when close_solver is implemented
-  // close_solver();
+  close_solver();
 }
 
 void GenericSolver::start_solver() {
-  assert(false);
+  pid = 0;
+
+  pipe(inpipefd);
+  pipe(outpipefd);
+  pid = fork();
+  if (pid == 0)
+  {
+    // Child
+    dup2(outpipefd[0], STDIN_FILENO);
+    dup2(inpipefd[1], STDOUT_FILENO);
+    dup2(inpipefd[1], STDERR_FILENO);
+
+    // ask kernel to deliver SIGTERM in case the parent dies
+    prctl(PR_SET_PDEATHSIG, SIGTERM);
+    //    fcntl(inpipefd[1], F_SETFL, O_NONBLOCK);
+    // The following part is based on:
+    // https://stackoverflow.com/a/5797901/1364765 The execv command expects an
+    // array, and so we create one. First element is the program name, last
+    // element is NULL, and in between are the elements of cmd_line_args, casted
+    // to (char*) from std::string. Here we identify the program name with its
+    // path.
+    const char ** argv = new const char *[cmd_line_args.size() + 2];
+    argv[0] = path.c_str();
+    for (int i = 1; i <= cmd_line_args.size(); i++)
+    {
+      argv[i] = cmd_line_args[i - 1].c_str();
+    }
+    argv[cmd_line_args.size() + 1] = NULL;
+    execv(path.c_str(), (char **)argv);
+    // Nothing below this line should be executed by child process. If so,
+    // it means that the execl function wasn't successfull, so lets exit:
+    string msg("failure to run binary: ");
+    msg += path;
+    throw IncorrectUsageException(msg);
+    exit(1);
+  }
+  // close unused pipe ends
+  close(outpipefd[0]);
+  close(inpipefd[1]);
+  set_opt("print-success", "true");
+}
+
+void GenericSolver::write_internal(string str) const
+{
+  // track how many charas were written so far
+  uint written_chars = 0;
+  // continue writing  until entire str was written
+  while (written_chars < str.size())
+  {
+    // how many characters are there left to write
+    uint remaining = str.size() - written_chars;
+    // how many characters are we writing in this iteration
+    uint substr_size;
+    if (remaining < write_buf_size)
+    {
+      substr_size = remaining;
+    }
+    else
+    {
+      substr_size = write_buf_size;
+    }
+    // write
+    strcpy(write_buf, str.substr(written_chars, substr_size).c_str());
+    write(outpipefd[1], write_buf, substr_size);
+    written_chars += substr_size;
+  }
+}
+
+bool GenericSolver::is_done(int just_read, std::string result) const
+{
+  bool done = false;
+  int count = 0;
+  // if we didn't read anything now, the command is done executing
+  if (just_read == 0)
```

Comment:
This is fine but a little funny to me that there's a flag that can determine the output. Do you only need `just_read` for this and the assertion below? I think it's fine as-is just thought I'd bring it up for you to consider.